### PR TITLE
Add socks5 proxy

### PR DESCRIPTION
**Pull request by danog**

### CODE REVIEW

**Comment by xpader** [comment]
What's the status now? Can this be merged?

File: src/Https1TunnelConnector.php [hunk 1]
```diff
@@ -69,23 +69,20 @@ public function connect(SocketAddress|string $uri, ?ConnectContext $context = nu
...
-        if ((string)$serverSocket->getRemoteAddress() !== (string)$clientSocket->getLocalAddress()) {
-            goto retry; // someone else connected faster...
-        }
+            try {
+                $serverSocket = $server->accept();
+                $clientSocket = $clientSocketFuture->await();
+            } finally {
+                $server->close();
+            }
+        } while (!$serverSocket || (string) $serverSocket->getRemoteAddress() !== (string) $clientSocket->getLocalAddress());
```
Comment (kelunik) [inline src/Https1TunnelConnector.php:85]: Why this new `!$serverSocket` check?

File: src/Internal/TunnelSocket.php [hunk 2]
```diff
@@ -60,14 +61,20 @@ public function end(): void
...
+    /** @api */
     public function reference(): void
     {
+        \assert($this->localSocket instanceof ResourceStream);
+        \assert($this->remoteSocket instanceof ResourceStream);
         $this->localSocket->reference();
         $this->remoteSocket->reference();
     }
 
+    /** @api */
```
Comment (kelunik) [inline src/Internal/TunnelSocket.php:73]: Why these annotations?

Comment (danog) [reply src/Internal/TunnelSocket.php:73]: It's to mark this unused function as public API so psalm doesn't complain about unused code

File: src/Https1TunnelConnector.php (hunk 1 above)
Comment (danog) [reply src/Https1TunnelConnector.php:85]: To make psalm happy

**Review by kelunik — APPROVED** [review]

**Pull request merged**